homeassistant/components/backup/manager.py

```diff
@@ -0,0 +1,171 @@
+"""Backup manager for the Backup integration."""
+from __future__ import annotations
+
+from dataclasses import dataclass
+import hashlib
+import json
+from pathlib import Path
+import tarfile
+from tempfile import TemporaryDirectory
+
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.util import json as json_util
+from homeassistant.util.dt import now
+
+from .const import EXCLUDE_FROM_BACKUP, LOGGER, VERSION
+
+
+@dataclass
+class Backup:
+    """Backup class."""
+
+    slug: str
+    name: str
+    date: str
+    path: Path
+    size: float
+
+    def as_dict(self) -> dict:
+        """Return a dict representation of this backup."""
+        return {
+            "slug": self.slug,
+            "name": self.name,
+            "date": self.date,
+            "size": self.size,
+            "path": self.path.as_posix(),
+        }
+
+
+class BackupManager:
+    """Backup manager for the Backup integration."""
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize the backup manager."""
+        self._backups: dict[str, Backup] = {}
+        self.hass = hass
+        self.backup_dir = Path(hass.config.path("backups"))
+        self.backing_up = False
+
+    @property
+    def backups(self) -> dict[str, Backup]:
+        """Return a dict of backups."""
+        return self._backups
+
+    def load_backups(self) -> None:
+        """Load data of stored backup files."""
+        backups = {}
+
+        def _read_backup_info(backup_path: Path) -> None:
+            with tarfile.open(backup_path, "r:") as backup_file:
+                if data_file := backup_file.extractfile("./backup.json"):
+                    data = json.loads(data_file.read())
+                    backup = Backup(
+                        slug=data["slug"],
+                        name=data["name"],
+                        date=data["date"],
+                        path=backup_path,
+                        size=round(backup_path.stat().st_size / 1_048_576, 2),
+                    )
+                    backups[backup.slug] = backup
+
+        for backup_path in self.backup_dir.glob("*.tar"):
+            _read_backup_info(backup_path)
+
+        LOGGER.debug("Loaded %s backups", len(backups))
+        self._backups = backups
+
+    def get_backup(self, slug: str) -> Backup | None:
+        """Return a backup."""
+        return self._backups.get(slug)
+
+    def remove_backup(self, slug: str) -> None:
+        """Remove a backup."""
+        if (backup := self.get_backup(slug)) is None:
+            return
+        backup.path.unlink(missing_ok=True)
+        LOGGER.debug("Removed backup located at %s", backup.path)
+        self._backups.pop(slug, None)
+
+    async def generate_backup(self) -> Backup:
+        """Generate a backup."""
+        try:
+            if self.backing_up:
+                raise HomeAssistantError("Backup already in progress")
```

Comment:
You need to put this outside the try, or else you are going to set `self.backing_up = False` in the `finally` while you were not the one that set it.
```suggestion
        if self.backing_up:
            raise HomeAssistantError("Backup already in progress")

        try:
```

Reply:
right!